Add code for training neural networks and creating neural network ICs

chaitanyaBhave26 opened the pull request:
- Added ExodusNNTrainer -> Directory in moose/python/
  - Added ExodusReader.py script to parse Exodus output files and generate training data for Pytorch neural networks
  - Added NeuralNet_to_XML.py script to convert PyTorch models into an XML tree that can be parsed in MOOSE using a 
    NeuralNetworkUserObject
- Added a NeuralNetworkUserObject to the stochastic_tools module
  - Parses XML tree to  recreate the PyTorch neural network model within MOOSE
  - Evaluates the neural network for an input DenseVector supplied by NeuralNetworkIC
- Added a NeuralNetworkIC to the stochastic_tools module
  - Reads variable values and generates a DenseVector that is fed to the NeuralNetworkUserObject to provide IC values for the variable
- Adds documentation for both objects added to stochastic_tools
- Adds example PyTorch training script that uses the python modules, along with input files that demonstrate the created objects
- Adds a tutorial explaining how the PyTorch script and NeuralNetwork objects are used. 

Closes: #15754

chaitanyaBhave26 committed:
Resolved git issues.

chaitanyaBhave26 committed:
Updated NeuralNetworkIC.md documentation

chaitanyaBhave26 committed:
Python script for parsing exodus files into training data
Python script for converting PyTorch neural networks into XML trees

chaitanyaBhave26 committed:
Cleaned up example input files.

dschwen committed:
Merge branch 'next' into nn_ic_PR

dschwen commented on modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/neural_net_inversion.md:
```diff
@@ -0,0 +1,55 @@
+# Using a neural network IC
+
+This example goes over how to train a neural network to invert concentration from chemical potential, for creating a sub-concentration IC for the KKS model. We use the [NeuralNetworkUserObject.md] and [NeuralNetworkIC.md].
```
```suggestion
This example goes over how to train a neural network to invert concentration from chemical potential, for creating a sub-concentration IC for the KKS model. We use the [NeuralNetworkUserObject](NeuralNetworkUserObject.md) and [NeuralNetworkIC](NeuralNetworkIC.md).
```

dschwen commented on modules/stochastic_tools/src/userobjects/NeuralNetworkUserObject.C:
```diff
@@ -0,0 +1,193 @@
+
+  ifile.open(_weightsFile);
+  if (!ifile)
+    {
```
all of this needs to be `clang-format`ed

dschwen commented on python/ExodusNNTrainer/NeuralNet_to_XML.py:
```diff
@@ -0,0 +1,71 @@
+import torch
```
neads a header with hash-bang and comments

aeslaughter replied:
I would like these files to be in a python directory: python/stochastic_tools

These files also need unit testing of their own

dschwen commented on python/ExodusNNTrainer/exodusReader.py:
```diff
@@ -0,0 +1,96 @@
+import chigger
```
same

aeslaughter commented:
Can you put all media files into the idaholab/large_media repository, it is my intention to move all media out of the repo to keep it all in once place.

aeslaughter commented on modules/stochastic_tools/include/userobjects/NeuralNetworkUserObject.h:
```diff
@@ -0,0 +1,45 @@
+// forward declarations
+class NeuralNetworkUserObject;
+
+template <>
+InputParameters validParams<GeneralUserObject>();
```
This can be removed and all of methods and members of this class that are not overrides need documentation

dschwen replied:
also remove unused forward declaraion

aeslaughter commented on modules/stochastic_tools/src/userobjects/NeuralNetworkUserObject.C:
```diff
@@ -0,0 +1,193 @@
+    {
+      paramError("Cannot read neural net file");
+    }
+  pugi::xml_parse_result result = doc.load(ifile);
```
This is likely going away, we are moving to JSON format

dschwen replied:
@aeslaughter in the interest of not letting this PR die once Chaitanya's internship ends (pretty soon), I suggest going with XML in this PR and transitioning over to JSON in a separate PR. I have interest in that as well and have some code that I will combine with Chaitanya's stuff.

aeslaughter replied:
That is fine. The biggest thing I would like done then is using python/stochastic_tools and at least have a few tests to make sure these tools run.

chaitanyaBhave26 committed:
Update modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/neural_net_inversion.md

Closes issue #15754

Co-authored-by: Daniel Schwen <daniel@schwen.de>

chaitanyaBhave26 committed:
1. Clang-formatted CPP files
2. Moved Python code to stochastic_tools/python
3. Moved example requiring phase_field to combined/examples/

Resolves issue #15754

dschwen commented on modules/stochastic_tools/include/userobjects/NeuralNetworkUserObject.h:
```diff
@@ -0,0 +1,45 @@
+  std::vector<DenseVector<Real>> _bias;
+
+  MultiMooseEnum _layerActivationFunctionEnum;
+
```
remove empty lines before closing braces and after opening braces

dschwen commented on modules/stochastic_tools/src/userobjects/NeuralNetworkUserObject.C:
```diff
@@ -0,0 +1,193 @@
+    }
+
+pugi::xml_node root = doc.document_element();
+std::string activation_layers;
```
this all does not look clang-formatted to me...

dschwen commented on modules/stochastic_tools/src/userobjects/NeuralNetworkUserObject.C:
```diff
@@ -0,0 +1,186 @@
+NeuralNetworkUserObject::NeuralNetworkUserObject(const InputParameters & parameters)
+  : GeneralUserObject(parameters),
+    _weightsFile(getParam<FileName>("weights_file")),
+    _layerActivationFunctionEnum(MultiMooseEnum("SIGMOID SOFTSIGN TANH LINEAR LOGSIGMOID"))
```
This should be an `std::vector` of an `enum class`. MultiMooseEnum is for capturing input parameter values.

dschwen commented on modules/stochastic_tools/src/userobjects/NeuralNetworkUserObject.C:
```diff
@@ -0,0 +1,186 @@
+    output(i) += _bias[n](i);
+  }
+
+  return (output(op_id));
```
remove extra parenthesis

dschwen commented on modules/stochastic_tools/test/tests/ics/neural_net_ic/evaluate.i:
```diff
@@ -0,0 +1,32 @@
+# [StochasticTools]
+# []
+#
+# [Samplers]
+#   [test]
+#     type = CartesianProduct
+#     linear_space_items = '0.25 1 10
+#                           0.25 1 10
+#                           0.25 1 10'
+#   []
+# []
+#
+# [VectorPostprocessors]
+#   [results]
+#     type = SurrogateTester
+#     model = surrogate
+#     sampler = test
+#     execute_on = final
+#   []
+# []
+
```
remove commented out stuff. doe this produce anything without an executioner block?!

chaitanyaBhave26 replied:
This input file is meant to run the neural_net_ic.i (another input file in same directory). It compares the result of the nn_ic simulation to a gold exodus file, to confirm the neural net is evaluated correctly and applied to the ic variable.

dschwen commented on modules/stochastic_tools/test/tests/ics/neural_net_ic/neural_net_ic.i:
```diff
@@ -0,0 +1,118 @@
+[]
+
+[Variables]
+
```
remove empty line

dschwen commented on modules/stochastic_tools/test/tests/ics/neural_net_ic/neural_net_ic.i:
```diff
@@ -0,0 +1,118 @@
+  [../]
+  [./w_Cr]
+  [../]
+
```
this one, too. By the way, it'd be great if you could run your inputs through `hit format` (there also is an Atom plugin for that)

dschwen commented on modules/stochastic_tools/test/tests/ics/neural_net_ic/tests:
```diff
@@ -0,0 +1,10 @@
+[Tests]
+#  issues = "#15361"
```
this will need a proper issue number or it will fail our SQA testing!

chaitanyaBhave26 committed:
1. Removed comment code from input files
2. Hit format applied to input files
3. Forward declarations removed from NeuralNetworkUserObject.h
4. Removed outdated setWeights function
5. Added some documentation lines to the CPP code

Resolves issue  #15754

chaitanyaBhave26 committed:
1. Added mooseError for softsign, will be implemented in future
2. Removed extra paranthesis in NeuralNetworkUserObject.C

Resolves issue  #15754

chaitanyaBhave26 committed:
1. Added issue # to tests file in neural_net_ic test

Resolves issue  #15754

chaitanyaBhave26 committed:
1. Fixed unused parameter warning in NeuralNetworkIC

Resolves #15754

chaitanyaBhave26 committed:
Removed media folder from stochastic modules.
Still need to commit the image file to large_media and update the documentation md file to reflect this change.
Resolves #15754

chaitanyaBhave26 committed:
1. Removed empty lines from neural_net_ic.i
2. Updated documentation for NN_IC example
3. Moved example to the correct location
Resolves  #15754
